lua/lazyvim/plugins/extras/lang/c.lua

```diff
@@ -0,0 +1,97 @@
+return {
+
+  -- Add C/C++ to treesitter
+  {
+    "nvim-treesitter/nvim-treesitter",
+    opts = function(_, opts)
+      if type(opts.ensure_installed) == "table" then
+        vim.list_extend(opts.ensure_installed, { "c", "cpp" })
+      end
+    end,
+  },
+
+  -- Correctly setup lspconfig for clangd 🚀
+  {
+    "neovim/nvim-lspconfig",
+    dependencies = {
+      "p00f/clangd_extensions.nvim",
+    },
+    opts = {
+      servers = {
+        -- Ensure mason installs the server
+        clangd = {
+          root_dir = function(...)
+            -- using a root .clang-format/.clang-tidy files mess up projects, so remove them
+            return require("lspconfig.util").root_pattern(
+              "compile_commands.json",
+              "compile_flags.txt",
+              "configure.ac",
+              ".git"
+            )(...)
+          end,
+        },
+      },
+      setup = {
+        clangd = function(_, opts)
+          opts.capabilities.offsetEncoding = { "utf-16" }
+          opts.cmd = {
+            "clangd",
+            "--background-index",
+            "--clang-tidy",
+            "--header-insertion=iwyu",
+            "--completion-style=detailed",
+            "--function-arg-placeholders",
+            "--fallback-style=llvm",
+          }
+          opts.init_options = {
+            usePlaceholders = true,
+            completeUnimported = true,
+            clangdFileStatus = true,
+          }
+          require("clangd_extensions").setup({
```

Comment:
so move these options to `opts` on line 12.

Then here, you just do `require("clangd_extensions")`
This will properly load and setup `clangd_extensions` and will use any custom options the user may have set.

Reply:
Hmm, maybe I'm overthinking this but clangd_extensions sets up the clangd lsp (hence the return true), so don't we need to call setup here (and if so - then how can we get the opts that the user has set for clangd inside here?)

Reply:
Have a spec outside of lspconfig, so not in its dep  like:
```lua
{".../extenstions", opts= {the options you currently pass to seup}, lazy = true}
```

Having that spec, users can easily override its `opts`.

Then where you currently call `extenstions.setup`, just do `require("clangd_extensions")` instead.
By doing this, lazy will see the plugin has not been loaded yet, and will automatically call `require("clangd_extensions").setup(ops)` where opts would be the mergded opts for the plugin.

We should do the same for rust-tools btw. Currently there it's done by using an internal lazy API, but the approach I suggest is much easier.

Reply:
Yeah that makes sense - thanks for the clarification. My only lingering question is how we can set the server key to be the clangd lspconfig opts `server = opts`. It's epecially important that offsetEncoding is utf-16 otherwise the user is bombarded with warning messages about multiple client encodings)

Reply:
Right! Didn't look at it that well it seems :)

Then the approach from https://github.com/LazyVim/LazyVim/blob/fca09b12e8188fb0a9b7bda040890c096e6cc9be/lua/lazyvim/plugins/extras/lang/rust.lua#L73 should be used instead.

But I'd still move all the opts to the opts of the plugin as I mentioned, but then get it's opts and call setup where you already did it.

Reply:
And make sure to move it outside of lspconfig deps with lazy=false.
Also add a `config = function()end` to the extensions spec to prevent it being loaded twice

Reply:
done..but should lazy be true or false? (you mentioned true at first, but false here)

Reply:
Sorry, lazy should be true indeed, to prevent it loading during startup

Reply:
done